### DISCUSSION TITLE
Tokio 1.0 Signal no longer implements Stream

### DISCUSSION THREAD

**Discussion by djmarcin** [tokio-rs/tokio#3432] [Q&A] [answered]
In tokio [0.3.6](https://docs.rs/tokio/0.3.6/tokio/signal/unix/struct.Signal.html#impl-Stream) tokio::signal::unix::Signal implemented Stream, but in [1.0.2](https://docs.rs/tokio/1.0.2/tokio/signal/unix/struct.Signal.html) it does not. This means it can no longer be passed to things like `stream::select_all`.

How should such code be migrated? Is there a way to wrap a Signal in a Stream?

**Comment by Darksonn** [comment, marked as answer]
You can turn it into a `Stream` using the `async-stream` crate.
```Rust
let stream = async_stream::stream! {
    loop {
        signal.recv().await;
        yield ();
    }
};
```

**Reply by djmarcin** [reply]
It looks like other things also don't implement stream anymore, like `tokio::sync::mpsc::Receiver`. Is the suggestion that all lines like:
```
while let Some(element) = rx.next().await {
  ...
}
```
get converted to:
```
let rx_stream = async_stream::stream! {
    loop {
        yield rx.recv().await;
    }
};
while let Some(element) = rx_stream.next().await {
  ...
}
```

**Reply by Darksonn** [reply]
In some cases you can use [`tokio_stream::wrappers`](https://docs.rs/tokio-stream/0.1/tokio_stream/wrappers/index.html). Though in your example, you would convert it to calling `.recv()`.

**Reply by Darksonn** [reply]
If you _did_ want to use async-stream for mpsc, it would look like this:
```Rust
let stream = async_stream::stream! {
    while let Some(item) = rx.recv().await {
        yield item;
    }
};
```
unlike your solution, this handles the channle being closed correctly.

**Comment by Roguelazer** [comment]
Are there any plans to add something to `tokio_stream::wrappers` that implements Stream for Signal? Would such an addition be an acceptable PR? Using `async_stream` is problematic because the macro interferes with `tokio::select!` (and generally complicates code quite a lot).

**Reply by Darksonn** [reply]
We already have #3383 open, which is necessary for writing the wrapper. If you are just using it in `select!`, why not just call `.recv()` on the `Signal`?
